Proposed Changes to Specification - Revision 1

vfscalfani opened the pull request:
This pull request is the first major proposed revision to the IUPAC SMILES+ specification. It is a work in progress and is still under review by the IUPAC SMILES+ Task Group (i.e., this pull request may change considerably). Below is a summary of the proposed changes and additions at the time of this writing:

1. Various typographic edits, updating of links/references, and added a purpose/motivation for the IUPAC SMILES+ specification, which build open the OpenSMILES work. In that section, I highlighted the importance of interoperability as a primary goal.

2. Proposed Extensions have been moved to a separate appendix, and an abridged version of the SMILES flavors section describing canonical terminology was merged into main text.

3. Atoms Section: Added support for element symbols through Og and element numbers in brackets through [#118], with [#0] as undefined. Also added support for aromatic `te`.

4. Hydrogen Section: Edited allowed Hydrogen `Hn`, as `n` being a single digit number from 0-9.

5. Charge Section: clarified atom charge property syntax and added support for repeated symbols up to 15 and a 0 charge. In the newly added syntax, for example, `[Ag+0]` and `[Ag-0]` are equivalent to `[Ag]`. Leading zeros are not allowed. Added that the charge symbol `'-'` or `'+'` must come before the numeric digit. That is, `[Mg+2]` is valid, while `[Mg2+]` is invalid. Moreover, providing multiple charge specifications such as `[Li+1-2]` or `[Li+-]` are invalid.

6. Isotope Section: Changed the interpretation of a zero isotope. Now, a zero isotope is equivalent to undefined. So, `[0S]` is equivalent to `[S]`. Further, leading zeros are now invalid (e.g., `[098Ru]` is invalid, it should be written as `[98Ru]`. Finally, the use of `D` or `T` symbols allowed in relaxed OpenSMILES parsing has been removed.

7. Wilidcard Section: Clarified that if `[#0]` notation is used, that `#0` can only appear within square brackets, in contrast to `*` which is part of the special organic subset, not requiring brackets. 

8. Atom Class: Atom class can no longer contain leading zeros.

9. Incorporated Nonstandard forms of SMILES section into main text (these were permitted with relaxed parsing). These are now considered invalid. Examples include:

* Extra parentheses such as `((C))` or `(N1CCCC1)`
* Misplaced dots such as `[Na+]..[Cl-]` and `.CCO`
* Mismatched ring bonds such as `C1CCC` or `C1CCCCC2`
* Invalid or conflicting cis/trans notation such as `C/C=C` and `C/C(\F)=C/C`
* Lowercase as sp2 outside of ring systems.

I also added some other examples of invalid SMILES into the main text such as duplicate bonds:
`CC=` and `C==C`, additional misplaced parentheses: `C(1CC1)`, `C()`, and `C1CC(1)`, and duplicate up/down specifications such as `C(/Br)=C\\I`.

10. Rings: Leading zeros in two-digit notation `%nn` is now considered invalid. Added support for three digit %(nnn) notation. Lastly, clarified parsing behavior when there is a conflicting ring closure bond: if there are conflicting ring closure bonds defined such as `C=1CCCCC#1`, `C-1CCCCC=1`, or `c=1ccccc:1`, the SMILES are invalid. Two defined bonds must match. In contrast, if only one bond is defined as in `C=1CCCCC1` or `C1CCCC=1`, the SMILES are valid and interpreted with the defined bond at ring closure. The implicit bond is ignored.

11. A strict bracket atom property order in brackets was emphasized. The valid order from left to right is isotope, atom symbol/number, chirality, hydrogen count, charge, then atom class.

12. Added that leading spaces in SMILES are invalid as well as empty strings.

vfscalfani committed:
removed blank notes file. Notes will now be maintained in PRs.

vfscalfani committed:
moved proposed extensions to a separate appendix

vfscalfani committed:
updated links and references

vfscalfani committed:
fixed reference lines.

vfscalfani committed:
update to purpose and motiv., aromatic te, element numbers, and H digits

vfscalfani committed:
symbol formatting fixes

vfscalfani committed:
asterisk formatting fixes

vfscalfani committed:
table formatting

vfscalfani committed:
general editing for consistency

vfscalfani committed:
edits to charge specification.

vfscalfani committed:
isotope spec changes

vfscalfani committed:
add edits to wildcard atom section.

vfscalfani committed:
minor formatting fix

vfscalfani committed:
more asciidoc formatting fixes....

vfscalfani committed:
removed lowercase as sp2

vfscalfani committed:
various changes noted in document rev history 9-29.

vfscalfani committed:
formatting fixes

vfscalfani committed:
removed old link

vfscalfani committed:
reformatted revision history table.

vfscalfani committed:
revision table size adjustment.

baoilleach commented:
What SMILES string should a program write for a molecule without any atoms, if the empty string is invalid?

vfscalfani commented:
Hi @baoilleach. I see what you mean, it's a good question. Is there a real use-case to write a SMILES string without any atoms? When an InChI string is empty, I have seen it written as `InChI=1S//` or an empty string `' '`. So, yeah, we'll need to think about this more. Thanks.

baoilleach commented:
Let's say you want to do an InChI -> SMILES conversion. If you have a file with 1000 InChIs, you want to have 1000 SMILES come out, but what do you do when the InChI fails to be read? Users weren't happy with 999 SMILES coming out - they couldn't work out which of the 1000 InChIs had the problem.
Another example is in reaction SMILES. Here's a failed reaction, "C>O>", and here's a reaction with no catalyst, "C>>O". Are these invalid for using the empty string for the product and catalyst?

vfscalfani commented:
We should also take a look at hydrogen n count again before merging. See this Issue https://github.com/openbabel/openbabel/issues/2344.

merkys commented on IUPAC_SMILES+.asciidoc:
```diff
@@ -102,7 +109,7 @@ http://www.daylight.com/meetings/mug03/McLeod/MUG03McLeodPeters.pdf[McLeod Peter
 
 McLeod and Peters' quip captures the deficiencies of SMILES well: if you
-*can't* build a molecule from a modeling kit, the deficiencies of SMILES and other
+*can not* build a molecule from a modeling kit, the deficiencies of SMILES and other
```
I am not a native English speaker, but ['cannot' seems to be more common form](https://en.wiktionary.org/wiki/cannot#Usage_notes). Moreover, the specification document already uses 'cannot' form.

merkys commented on IUPAC_SMILES+.asciidoc:
```diff
@@ -212,23 +227,21 @@ Examples:
-For backwards compatibility, a general-purpose SMILES parser should
-accept the symbols `--` and `++` to mean charges of `-2` and `+2`, but this
-is a deprecated form and should be avoided.
+Repeated symbols such as `'--'` and `'++'` are valid and interpreted as charges of `-2` and `+2`, respectively. Symbols can
```
Repeats of more than two symbols are not covered by the grammar, thus statement "Symbols can repeat up to 15 times" seems to contradict the grammar. As allowing more repeats is intentional (described in the changelog), grammar should be updated to reflect it.

merkys commented on IUPAC_SMILES+.asciidoc:
```diff
@@ -320,37 +330,55 @@ Examples:
+meaningful only to the application. Multiple atoms may be labeled with the same atom class.
+
+If the atom class is not specified, then the atom class is interpreted as zero.
+The atom class can not contain a leading `0`. So for example, `[NH4+:5]` is valid, 
```
Same comment about 'can not' as above.

merkys commented on IUPAC_SMILES+.asciidoc:
```diff
@@ -1125,38 +1170,36 @@ Very few SMILES systems actually implement the rules for `SP`, `TB` or `OH` chir
 
 A SMILES string is terminated by a whitespace terminator character (space, tab, newline,
-carriage-return), or by the end of the string.
+carriage-return), or by the end of the string. As a result, and leading space in a SMILES string
```
I believe "As a result, and leading space" should read "As a result, any leading space".

merkys commented on IUPAC_SMILES+.asciidoc:
```diff
@@ -1125,38 +1170,36 @@ Very few SMILES systems actually implement the rules for `SP`, `TB` or `OH` chir
-100 ring-closure numbers, but since numbers can be reused, a molecule can potentially have more
-than 100 rings.  SMILES parsers should accept and correctly parse molecules with at least 1000 rings;
+There is no formal limit to the number of rings a molecule can contain. There are only
+999 ring-closure numbers, but since numbers can be reused, a molecule can potentially have more
```
Since '0' is a valid ring number, I believe there are 1000 ring-closure numbers, or am I wrong?

merkys commented on IUPAC_SMILES+.asciidoc:
```diff
@@ -1331,6 +1378,14 @@ thus changing the SMILES it produces.  A Canonical SMILES is primarily
 useful in a single database, or a system of related databases or information,
 in which all molecules were created using a single canonicalizer.
 
+It is an unfortunately common misconception that a Canonical SMILES does not (or can not)
```
Same comment on 'cannot' as above.

vfscalfani committed:
fixed typos, updated dates, and edits based on PR comments.

vfscalfani commented:
I am going to merge this now. I made a few changes based on comments:

1. removed that an empty string is invalid. I created an issue for this and will revisit it later.
2. fixed a few typos and updated dates.

@merkys 

1. I standardized the language as "can not", I have always been taught that "cannot" is informal. I can change this if you feel strongly about that. 
2. I updated the formal grammar table with the repeated charges, thanks.
3. Updated valid ring numbers. That's correct, so it is 0, and 1-999 =  1000.

If you have other comments, feel free to open a new issue.

vfscalfani committed:
fixed a typo with total number of rings.

vfscalfani committed:
fixed the plus rendering in formal grammar table
